flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogMigrationRestoreTarget.java

```diff
@@ -0,0 +1,223 @@
+                    throws Exception {
+                return keyedStateBackend.createInternalState(
+                        namespaceSerializer, stateDesc, snapshotTransformFactory);
+            }
+
+            @Override
+            public <N, S extends State> S getPartitionedState(
+                    N namespace,
+                    TypeSerializer<N> namespaceSerializer,
+                    StateDescriptor<S, ?> stateDescriptor)
+                    throws Exception {
+                S partitionedState =
+                        keyedStateBackend.getPartitionedState(
+                                namespace, namespaceSerializer, stateDescriptor);
+                functionDelegationHelper.addOrUpdate(stateDescriptor);
+                return partitionedState;
+            }
+
+            @Override
+            public <N, S extends State, V> S getOrCreateKeyedState(
+                    TypeSerializer<N> namespaceSerializer, StateDescriptor<S, V> stateDescriptor)
+                    throws Exception {
+                S keyedState =
+                        keyedStateBackend.getOrCreateKeyedState(
+                                namespaceSerializer, stateDescriptor);
+                functionDelegationHelper.addOrUpdate(stateDescriptor);
+                return keyedState;
+            }
+
+            @Nonnull
+            @Override
+            @SuppressWarnings("unchecked")
+            public <T extends HeapPriorityQueueElement & PriorityComparable<? super T> & Keyed<?>>
+                    KeyGroupedInternalPriorityQueue<T> create(
+                            @Nonnull String stateName,
+                            @Nonnull TypeSerializer<T> byteOrderedElementSerializer) {
+                ChangelogKeyGroupedPriorityQueue<T> existingState =
+                        (ChangelogKeyGroupedPriorityQueue<T>)
+                                changelogStateFactory.getExistingState(
+                                        stateName,
+                                        StateMetaInfoSnapshot.BackendStateType.PRIORITY_QUEUE);
+                return existingState == null
+                        ? keyedStateBackend.create(stateName, byteOrderedElementSerializer)
+                        : existingState;
+            }
+
+            @Nonnull
+            @Override
+            public RunnableFuture<SnapshotResult<KeyedStateHandle>> snapshot(
+                    long checkpointId,
+                    long timestamp,
+                    @Nonnull CheckpointStreamFactory streamFactory,
+                    @Nonnull CheckpointOptions checkpointOptions)
+                    throws Exception {
+                return keyedStateBackend.snapshot(
+                        checkpointId, timestamp, streamFactory, checkpointOptions);
+            }
+
+            @Override
+            public void dispose() {
+                super.dispose();
+                changelogStateFactory.dispose();
```

Comment:
Why it is not `keyedStateBackend.dispose()`, by analogy with other methods?

Reply:
You are right. Just modified.